lib/client/identityfile/identity.go

```diff
@@ -614,10 +614,15 @@ func KeyFromIdentityFile(identityPath, proxyHost, clusterName string) (*client.K
 // is also added to its profile store with the limited profile data available
 // in the identity file.
 //
-// Since identity files do not save a proxy address, proxyAddr must be provided
-// to fill in this data gap. clusterName can also be provided to aim the key at
-// a leaf cluster rather than the default root cluster.
+// proxyAddr: Since identity files do not save a proxy address, a host:port-like
+// proxyAddr must be provided to fill in this data gap.
+//
+// clusterName: may optionally be provided to aim the key at a leaf cluster
+// rather than the default root cluster.
 func NewClientStoreFromIdentityFile(identityFile, proxyAddr, clusterName string) (*client.Store, error) {
+	if proxyAddr == "" {
+		return nil, trace.BadParameter("missing proxy address. Specify --proxy or login with `tsh login`")
```

Comment:
I'm not sure it makes sense to mention specific CLI flags in an error message this deep. This function is not required to be called from a CLI, so we may get confusing error messages if this is mistakenly raised in the wrong context.

Your previous implementation was better IMO, because you mentioned the `--proxy` flag in a func that accepted a struct with command line flags.

Reply:
This one is a bit trickier to resolve, as it seems like there are testing and other usages of tool/tsh/tsh.go/`makeClient` which rely on `(CLIConf)cf.Proxy` being unset (and I guess it gets discovered/set in other ways in those contexts?), so putting this check up in `makeClient` also seems like it'll impact other non-CLI uses of tsh.

At least my thinking here is that only in this `NewClientStoreFromIdentityFile()` context are we in this one context we care about: building a tsh client up from just an identity file.

However, maybe you are right that there are other non-CLI contexts that this tsh Client could be getting built up from an identity file.
Is there some elegant way that we could detect the CLI-usage context and build up the ideal error message (but still here)?

Reply:
Maybe a better place for this check is up inside of `Run()` where most of the argument parsing is happening?

I would check up there, but I'm not confident about creating a conditional that could cover all the cases that might eventually try and construct the client from an identity file.

Intuitively, it make sense to put something like `if cf.IdentityFileIn && !(cf.Proxy)` up there, but I wasn't confident about how proxy discovery from the `~/.tsh/*` might interact with this.

Might it be possible that somebody has logged into Teleport, has valid proxy addresses via `~/.tsh/current-profile`, but then also calls `tsh` with an Identity file, such that they're not passing a proxy on the CLI, but they have a proxy address that could be discovered via the local state to use with the identity file?

Reply:
After testing some more cases with local config, this seems even more straightforward than I had initially thought.

I've decided to go with both some conditional checks for this condition when arguments are parsed as well as a check down in `NewClientStoreFromIdentityFile` to give a better error message if `proxyAddr` is empty.